File: packages/capacitor-on-device-briefing/android/src/main/java/app/gomsinlog/ondevicebriefing/OnDeviceBriefingPlugin.kt
package app.gomsinlog.ondevicebriefing

import android.os.Build
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONObject

/**
 * Capacitor bridge for On-Device Partner Briefing selection on Android.
 *
 * Bridge Name: GomsinlogOnDeviceBriefing
 * Methods:
 * - availability
 * - capability
 * - selectExtracts
 * - cancel
 */
@CapacitorPlugin(name = "GomsinlogOnDeviceBriefing")
class OnDeviceBriefingPlugin : Plugin() {

    private val pluginScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val engine by lazy { OnDeviceBriefingEngine(pluginScope, context) }

    @PluginMethod
    fun availability(call: PluginCall) {
        val keys = call.data.keys().asSequence().toSet()
        if (keys != setOf("locale")) {
            reject(call, BriefingErrorCode.BAD_REQUEST)
            return
        }

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            val ret = JSObject()
            ret.put("availability", OnDeviceBriefingAvailability.UNSUPPORTED.value)
            call.resolve(ret)
            return
        }

        val locale = call.getString("locale") ?: ""
        pluginScope.launch {
            val status = engine.checkAvailability(locale)
            val ret = JSObject()
            ret.put("availability", status.value)
            call.resolve(ret)
        }
    }

    @PluginMethod
    fun capability(call: PluginCall) {
        val keys = call.data.keys().asSequence().toSet()
        if (keys.isNotEmpty()) {
            reject(call, BriefingErrorCode.BAD_REQUEST)
            return
        }

        val envelope = JSObject().apply {
            put("maxContextUtf8Bytes", OnDeviceBriefing.MAX_CONTEXT_UTF8_BYTES)
            put("promptOverheadUtf8Bytes", OnDeviceBriefing.PROMPT_OVERHEAD_UTF8_BYTES)
            put("responseReserveUtf8Bytes", OnDeviceBriefing.RESPONSE_RESERVE_UTF8_BYTES)
            put("maxInputTextGraphemes", OnDeviceBriefing.MAX_INPUT_TEXT_GRAPHEMES)
            // Structural limits this parser already enforces. Advertised so the JS
            // batcher stops building requests this plugin will reject outright.
            put("maxItems", OnDeviceBriefing.MAX_ITEMS)
            put("maxCandidatesPerItem", OnDeviceBriefing.MAX_CANDIDATES_PER_ITEM)
        }
        val ret = JSObject().apply {
            put("envelope", envelope)
        }
        call.resolve(ret)
    }

    @PluginMethod
    fun selectExtracts(call: PluginCall) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            reject(call, BriefingErrorCode.UNAVAILABLE)
            return
        }

        val keys = call.data.keys().asSequence().toSet()
        if (keys != setOf("requestId", "locale", "items")) {
            reject(call, BriefingErrorCode.BAD_REQUEST)
            return
        }

        val requestId = call.getString("requestId")
        val locale = call.getString("locale")
        val itemsArray = call.getArray("items")

        if (requestId.isNullOrEmpty() ||
            requestId.toByteArray(Charsets.UTF_8).size > 128 ||
            locale.isNullOrEmpty() ||
            !OnDeviceBriefing.isLocaleSupported(locale) ||
            itemsArray == null
        ) {
            reject(call, BriefingErrorCode.BAD_REQUEST)
            return
        }

        val parsed = parseItems(itemsArray)
        if (parsed == null) {
            reject(call, BriefingErrorCode.BAD_REQUEST)
            return
        }

        val deferred = try {
            engine.startSelect(
                requestId,
                locale,
                parsed.items,
                parsed.jsonString
            )
        } catch (e: OnDeviceBriefingException) {
            reject(call, e.errorCode)
            return
        } catch (_: Throwable) {
            reject(call, BriefingErrorCode.NATIVE_ERROR)
            return
        }

        pluginScope.launch {
            try {
                val groups = deferred.await()

                val groupsJsArray = JSArray()
                for (group in groups) {
                    val choicesJsArray = JSArray()
                    for (choice in group.choices) {
                        val choiceObj = JSObject().apply {
                            put("itemOrdinal", choice.itemOrdinal)
                            put("candidateOrdinal", choice.candidateOrdinal)
                        }
                        choicesJsArray.put(choiceObj)
                    }
                    val groupObj = JSObject().apply {
                        put("groupOrdinal", group.groupOrdinal)
                        put("choices", choicesJsArray)
                    }
                    groupsJsArray.put(groupObj)
                }

                val outputObj = JSObject().apply {
                    put("version", 2)
                    put("groups", groupsJsArray)
                }

                val result = JSObject().apply {
                    put("requestId", requestId)
                    put("output", outputObj)
                }
                call.resolve(result)
            } catch (_: CancellationException) {
                reject(call, BriefingErrorCode.CANCELLED)
            } catch (e: OnDeviceBriefingException) {
                reject(call, e.errorCode)
            } catch (_: Throwable) {
                reject(call, BriefingErrorCode.NATIVE_ERROR)
            }
        }
    }

    @PluginMethod
    fun cancel(call: PluginCall) {
        val keys = call.data.keys().asSequence().toSet()
        if (keys != setOf("requestId")) {
            reject(call, BriefingErrorCode.BAD_REQUEST)
            return
        }

        val requestId = call.getString("requestId")
        if (requestId.isNullOrEmpty() || requestId.toByteArray(Charsets.UTF_8).size > 128) {
            reject(call, BriefingErrorCode.BAD_REQUEST)
            return
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            engine.cancel(requestId)
        }
        call.resolve(JSObject())
    }

    override fun handleOnDestroy() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            engine.cancelAll()
        }
        pluginScope.cancel()
        super.handleOnDestroy()
    }

    private data class ParsedItems(
        val items: List<BriefingItem>,
        val jsonString: String
    )

    private fun parseItems(rawItems: JSArray): ParsedItems? {
        if (rawItems.length() == 0 || rawItems.length() > OnDeviceBriefing.MAX_ITEMS) {
            return null
        }

        val items = ArrayList<BriefingItem>(rawItems.length())
        val jsonItems = JSONArray()
        var totalGraphemes = 0

        for (i in 0 until rawItems.length()) {
            val itemObj = rawItems.optJSONObject(i) ?: return null

            val keys = itemObj.keys().asSequence().toSet()
            if (keys != setOf("itemOrdinal", "candidates")) {
                return null
            }

            val itemOrdinalVal = itemObj.opt("itemOrdinal")
            if (itemOrdinalVal !is Int && (itemOrdinalVal !is Number || itemOrdinalVal.toDouble() != itemOrdinalVal.toInt().toDouble())) {
                return null
            }
            val itemOrdinal = (itemOrdinalVal as Number).toInt()
            if (itemOrdinal != i) {
                return null
            }

            val candidatesArray = itemObj.optJSONArray("candidates") ?: return null
            if (candidatesArray.length() == 0 || candidatesArray.length() > OnDeviceBriefing.MAX_CANDIDATES_PER_ITEM) {
                return null
            }

            val candidates = ArrayList<BriefingCandidate>(candidatesArray.length())
            val jsonCandidates = JSONArray()

            for (c in 0 until candidatesArray.length()) {
                val candidateObj = candidatesArray.optJSONObject(c) ?: return null
                val candidateKeys = candidateObj.keys().asSequence().toSet()
                if (candidateKeys != setOf("candidateOrdinal", "text")) {
                    return null
                }

                val candidateOrdinalVal = candidateObj.opt("candidateOrdinal")
                if (candidateOrdinalVal !is Int && (candidateOrdinalVal !is Number || candidateOrdinalVal.toDouble() != candidateOrdinalVal.toInt().toDouble())) {
                    return null
                }
                val candidateOrdinal = (candidateOrdinalVal as Number).toInt()
                if (candidateOrdinal != c) {
                    return null
                }

                val rawText = candidateObj.opt("text")
                if (rawText !is String) {
                    return null
                }
                val text = rawText
                if (text.trim().isEmpty()) {
                    return null
                }

                val graphemes = engine.countGraphemes(text)
                totalGraphemes += graphemes
                if (totalGraphemes > OnDeviceBriefing.MAX_INPUT_TEXT_GRAPHEMES) {
                    return null
                }

                candidates.add(BriefingCandidate(candidateOrdinal, text))
                val jsonCandidate = JSONObject().apply {
                    put("candidateOrdinal", candidateOrdinal)
                    put("text", text)
                }
                jsonCandidates.put(jsonCandidate)
            }

            items.add(BriefingItem(itemOrdinal, candidates))
            val jsonItem = JSONObject().apply {
                put("itemOrdinal", itemOrdinal)
                put("candidates", jsonCandidates)
            }
            jsonItems.put(jsonItem)
        }

        val jsonString = jsonItems.toString()
        val utf8Bytes = jsonString.toByteArray(Charsets.UTF_8).size
        if (utf8Bytes + OnDeviceBriefing.PROMPT_OVERHEAD_UTF8_BYTES + OnDeviceBriefing.RESPONSE_RESERVE_UTF8_BYTES > OnDeviceBriefing.MAX_CONTEXT_UTF8_BYTES) {
            return null
        }

        return ParsedItems(items, jsonString)
    }

    private fun reject(call: PluginCall, errorCode: BriefingErrorCode) {
        call.reject(errorCode.defaultMessage, errorCode.code)
    }
}

File: packages/capacitor-on-device-briefing/ios/Sources/OnDeviceBriefingPlugin/OnDeviceBriefingPlugin.swift
import Foundation
import Capacitor

@objc(GomsinlogOnDeviceBriefingPlugin)
public class GomsinlogOnDeviceBriefingPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "GomsinlogOnDeviceBriefingPlugin"
    public let jsName = "GomsinlogOnDeviceBriefing"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "availability", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "capability", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "selectExtracts", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancel", returnType: CAPPluginReturnPromise),
    ]

    @objc func availability(_ call: CAPPluginCall) {
        guard requireExactKeys(call, expected: ["locale"]) else {
            reject(call, error: .badRequest)
            return
        }
        let locale = call.getString("locale") ?? ""
        call.resolve([
            "availability": OnDeviceBriefing.availability(locale: locale).rawValue,
        ])
    }

    @objc func capability(_ call: CAPPluginCall) {
        guard requireExactKeys(call, expected: []) else {
            reject(call, error: .badRequest)
            return
        }
        call.resolve([
            "envelope": [
                "maxContextUtf8Bytes": OnDeviceBriefing.maxContextUtf8Bytes,
                "promptOverheadUtf8Bytes": OnDeviceBriefing.promptOverheadUtf8Bytes,
                "responseReserveUtf8Bytes": OnDeviceBriefing.responseReserveUtf8Bytes,
                "maxInputTextGraphemes": OnDeviceBriefing.maxInputTextGraphemes,
                // Structural limits this parser already enforces below. Advertised so the
                // JS batcher stops building requests this plugin will reject outright:
                // a record segmenting into 33 candidates was accepted by JS and refused
                // here, and the couple got deterministic output on a capable device.
                "maxItems": OnDeviceBriefing.maxItems,
                "maxCandidatesPerItem": OnDeviceBriefing.maxCandidatesPerItem,
            ],
        ])
    }

    @objc func selectExtracts(_ call: CAPPluginCall) {
        guard requireExactKeys(call, expected: ["requestId", "locale", "items"]) else {
            reject(call, error: .badRequest)
            return
        }
        guard let requestId = call.getString("requestId"),
              !requestId.isEmpty,
              requestId.utf8.count <= 128,
              let locale = call.getString("locale"),
              OnDeviceBriefing.localeIdentifier(for: locale) != nil,
              let parsed = parseItems(call) else {
            reject(call, error: .badRequest)
            return
        }

        Task {
            do {
                let groups = try await OnDeviceBriefingEngine.shared.select(
                    requestId: requestId,
                    locale: locale,
                    items: parsed.items,
                    itemsJSON: parsed.json
                )
                call.resolve([
                    "requestId": requestId,
                    "output": [
                        "version": 2,
                        "groups": groups.map { group in
                            [
                                "groupOrdinal": group.groupOrdinal,
                                "choices": group.choices.map { choice in
                                    [
                                        "itemOrdinal": choice.itemOrdinal,
                                        "candidateOrdinal": choice.candidateOrdinal,
                                    ]
                                },
                            ]
                        },
                    ],
                ])
            } catch is CancellationError {
                call.reject("on-device briefing was cancelled", "E_CANCELLED")
            } catch let error as OnDeviceBriefingError {
                reject(call, error: error)
            } catch {
                call.reject("on-device briefing failed", "E_NATIVE")
            }
        }
    }

    @objc func cancel(_ call: CAPPluginCall) {
        guard requireExactKeys(call, expected: ["requestId"]) else {
            reject(call, error: .badRequest)
            return
        }
        guard let requestId = call.getString("requestId"),
              !requestId.isEmpty,
              requestId.utf8.count <= 128 else {
            reject(call, error: .badRequest)
            return
        }
        Task {
            await OnDeviceBriefingEngine.shared.cancel(requestId: requestId)
            call.resolve([:])
        }
    }

    private func parseItems(
        _ call: CAPPluginCall
    ) -> (items: [OnDeviceBriefingItem], json: String)? {
        guard let rawItems = call.getArray("items"),
              !rawItems.isEmpty,
              rawItems.count <= OnDeviceBriefing.maxItems else {
            return nil
        }

        var parsed: [OnDeviceBriefingItem] = []
        var jsonItems: [[String: Any]] = []
        var totalGraphemes = 0

        for rawItem in rawItems {
            guard let entry = rawItem as? [String: Any],
                  Set(entry.keys) == Set(["itemOrdinal", "candidates"]),
                  let itemOrdinal = exactInteger(entry["itemOrdinal"]),
                  itemOrdinal == parsed.count,
                  let rawCandidates = entry["candidates"] as? [Any],
                  !rawCandidates.isEmpty,
                  rawCandidates.count <= OnDeviceBriefing.maxCandidatesPerItem else {
                return nil
            }

            var candidates: [OnDeviceBriefingCandidate] = []
            var jsonCandidates: [[String: Any]] = []
            for rawCandidate in rawCandidates {
                guard let candidate = rawCandidate as? [String: Any],
                      Set(candidate.keys) == Set(["candidateOrdinal", "text"]),
                      let candidateOrdinal = exactInteger(candidate["candidateOrdinal"]),
                      candidateOrdinal == candidates.count,
                      let text = candidate["text"] as? String,
                      !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                    return nil
                }
                totalGraphemes += text.count
                guard totalGraphemes <= OnDeviceBriefing.maxInputTextGraphemes else {
                    return nil
                }
                candidates.append(.init(candidateOrdinal: candidateOrdinal, text: text))
                jsonCandidates.append([
                    "candidateOrdinal": candidateOrdinal,
                    "text": text,
                ])
            }

            parsed.append(.init(itemOrdinal: itemOrdinal, candidates: candidates))
            jsonItems.append([
                "itemOrdinal": itemOrdinal,
                "candidates": jsonCandidates,
            ])
        }

        guard JSONSerialization.isValidJSONObject(jsonItems),
              let data = try? JSONSerialization.data(withJSONObject: jsonItems),
              data.count + OnDeviceBriefing.promptOverheadUtf8Bytes
                  + OnDeviceBriefing.responseReserveUtf8Bytes
                  <= OnDeviceBriefing.maxContextUtf8Bytes,
              let json = String(data: data, encoding: .utf8) else {
            return nil
        }
        return (parsed, json)
    }

    private func requireExactKeys(_ call: CAPPluginCall, expected: Set<String>) -> Bool {
        let actualKeys = Set(call.options.keys.compactMap { $0 as? String })
        return actualKeys == expected
    }

    private func exactInteger(_ value: Any?) -> Int? {
        if value is Bool { return nil }
        if let integer = value as? Int { return integer >= 0 ? integer : nil }
        guard let number = value as? NSNumber else { return nil }
        let integer = number.intValue
        return integer >= 0 && number.doubleValue == Double(integer) ? integer : nil
    }

    private func reject(_ call: CAPPluginCall, error: OnDeviceBriefingError) {
        switch error {
        case .unavailable:
            call.reject("on-device briefing unavailable", "E_UNAVAILABLE")
        case .badRequest:
            call.reject("request is outside the contract", "E_BAD_REQUEST")
        case .malformedOutput:
            call.reject("on-device briefing returned an unusable shape", "E_MALFORMED")
        case .busy:
            call.reject("on-device briefing is busy", "E_BUSY")
        case .quota:
            call.reject("on-device briefing quota unavailable", "E_QUOTA")
        case .nativeFailure:
            call.reject("on-device briefing failed", "E_NATIVE")
        }
    }
}
